File: src/main/java/me/testaccount666/serversystem/listener/executables/joinquitnotifier/ListenerQuit.java
package me.testaccount666.serversystem.listener.executables.joinquitnotifier;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.managers.config.ConfigurationManager;
import me.testaccount666.serversystem.managers.messages.MessageManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.userdata.UserManager;
import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.Registry;
import org.bukkit.Sound;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerQuitEvent;

public class ListenerQuit implements Listener {
    private final boolean _modifyMessage;
    private final boolean _sendMessage;
    private final String _message;
    private boolean _playSound;
    private Sound _sound;

    public ListenerQuit() {
        var configManager = ServerSystem.getInstance().getRegistry().getService(ConfigurationManager.class);
        var config = configManager.getGeneralConfig();
        _modifyMessage = config.getBoolean("Quit.Message.Enabled", false);
        _sendMessage = config.getBoolean("Quit.Message.SendMessage", false);
        _message = config.getString("Quit.Message.Message", "");
        _playSound = config.getBoolean("Quit.Sound.Enabled", false);
        if (!_playSound) {
            _sound = null;
            return;
        }

        var soundString = config.getString("Quit.Sound.Sound", "");

        var isMinecraft = true;
        if (soundString.contains(":")) {
            var space = soundString.substring(0, soundString.indexOf(":"));
            soundString = soundString.substring(soundString.indexOf(":") + 1);
            isMinecraft = space.equalsIgnoreCase("minecraft");
        }

        var soundKey = isMinecraft? NamespacedKey.minecraft(soundString) : NamespacedKey.fromString(soundString);
        if (soundKey == null) {
            _playSound = false;
            ServerSystem.getLog().warning("Failed to parse sound '${soundString}' for quit message!");
            return;
        }

        _sound = Registry.SOUND_EVENT.get(soundKey);
        if (_sound == null) {
            _playSound = false;
            ServerSystem.getLog().warning("Failed to find sound '${soundString}' for quit message!");
        }
    }

    @EventHandler
    public void onPlayerJoin(PlayerQuitEvent event) {
        handleMessage(event);
        playSound();
    }

    private void playSound() {
        if (!_playSound) return;

        Bukkit.getOnlinePlayers().forEach(everyone -> everyone.playSound(everyone, _sound, 1F, 1F));
    }

    private void handleMessage(PlayerQuitEvent event) {
        if (!_modifyMessage) return;

        if (!_sendMessage) {
            event.setQuitMessage(null);
            return;
        }
        var player = event.getPlayer();
        var userOptional = ServerSystem.getInstance().getRegistry().getService(UserManager.class).getUser(player);
        if (userOptional.isEmpty()) {
            ServerSystem.getLog().warning("Couldn't cache User '${player.getName()}'! This should not happen!");
            return;
        }
        var cachedUser = userOptional.get();
        if (cachedUser.isOfflineUser()) {
            ServerSystem.getLog().warning("User '${player.getName()}' is cached as Offline User! This should not happen!");
            return;
        }
        var user = (User) cachedUser.getOfflineUser();

        var message = MessageManager.formatMessage(_message, user, null, null, false);
        event.setQuitMessage(message);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/listener/executables/joinquitnotifier/ListenerQuit.kt
package me.testaccount666.serversystem.listener.executables.joinquitnotifier

import me.testaccount666.paperktx.extensions.ComponentExtensions.asComponent
import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.extensions.getService
import me.testaccount666.serversystem.managers.config.ConfigurationManager
import me.testaccount666.serversystem.managers.messages.MessageManager.formatMessage
import me.testaccount666.serversystem.userdata.UserManager
import org.bukkit.*
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent

class ListenerQuit : Listener {
    private val _modifyMessage: Boolean
    private val _sendMessage: Boolean
    private val _message: String
    private var _playSound: Boolean
    private var _sound: Sound? = null

    constructor() {
        val config = getService<ConfigurationManager>().generalConfig
        _modifyMessage = config.getBoolean("Quit.Message.Enabled")
        _sendMessage = config.getBoolean("Quit.Message.SendMessage")
        _message = config.getValue("Quit.Message.Message", "")
        _playSound = config.getBoolean("Quit.Sound.Enabled")
        if (!_playSound) {
            _sound = null
            return
        }

        var soundString = config.getValue("Quit.Sound.Sound", "")

        var isMinecraft = true
        if (soundString.contains(":")) {
            val space = soundString.substring(0, soundString.indexOf(":"))
            soundString = soundString.drop(soundString.indexOf(":") + 1)
            isMinecraft = space.contentEquals("minecraft", true)
        }

        val soundKey = (if (isMinecraft) NamespacedKey.minecraft(soundString)
        else NamespacedKey.fromString(soundString)) ?: run {
            _playSound = false
            log.warning("Failed to parse sound '${soundString}' for quit message!")
            return
        }

        _sound = Registry.SOUND_EVENT.get(soundKey) ?: let {
            _playSound = false
            log.warning("Failed to find sound '${soundString}' for quit message!")
            null
        }
    }

    @EventHandler
    fun onPlayerJoin(event: PlayerQuitEvent) {
        handleMessage(event)
        playSound()
    }

    private fun playSound() {
        if (!_playSound) return

        Bukkit.getOnlinePlayers().forEach { everyone: Player? -> everyone!!.playSound(everyone, _sound!!, 1f, 1f) }
    }

    private fun handleMessage(event: PlayerQuitEvent) {
        if (!_modifyMessage) return

        if (!_sendMessage) {
            event.quitMessage(null)
            return
        }
        val player = event.getPlayer()
        val user = getService<UserManager>().getUserOrNull(player) ?: run {
            log.warning("Couldn't cache User '${player.name}'! This should not happen!")
            return
        }
        if (user.isOfflineUser) {
            log.warning("User '${player.name}' is cached as Offline User! This should not happen!")
            return
        }
        formatMessage(_message, user.onlineUser, null, null, false).asComponent().also(event::quitMessage)
    }
}
